### PR TITLE
Fix TypeScript types for StrictHTMLCollection

### LINKED ISSUE #92: Typescript types for StrictHTMLFormElements are broken in 0.0.6
### Describe the issue

While trying to bump RSD to latest in our fluent repo (https://github.com/microsoft/fluentui/tree/xplat/apps/xplat-v9) I hit the current set of errors during the build step:

<details>
<summary>Error log</summary>

```
node_modules/react-strict-dom/dist/types/StrictHTMLFormElements.d.ts:39:3 - error TS1131: Property or signature expected.

39   @@iterator(): Iterator<Elem>;
     ~

node_modules/react-strict-dom/dist/types/StrictHTMLFormElements.d.ts:39:15 - error TS1109: Expression expected.

39   @@iterator(): Iterator<Elem>;


node_modules/react-strict-dom/dist/types/StrictHTMLFormElements.d.ts:40:3 - error TS1128: Declaration or statement expected.

40   readonly length: number;
     ~~~~~~~~

node_modules/react-strict-dom/dist/types/StrictHTMLFormElements.d.ts:41:18 - error TS1005: ']' expected.

41   readonly [index: number | string]: Elem;
                    ~

node_modules/react-strict-dom/dist/types/StrictHTMLFormElements.d.ts:41:35 - error TS1005: ';' expected.

41   readonly [index: number | string]: Elem;
                                     ~

node_modules/react-strict-dom/dist/types/StrictHTMLFormElements.d.ts:41:36 - error TS1128: Declaration or statement expected.

41   readonly [index: number | string]: Elem;
                                      ~

node_modules/react-strict-dom/dist/types/StrictHTMLFormElements.d.ts:42:1 - error TS1128: Declaration or statement expected.

42 }
   ~


Found 7 errors in the same file, starting at: node_modules/react-strict-dom/dist/types/StrictHTMLFormElements.d.ts:39
```

</details>

After some digging, it seems that this is caused by this new `StrictHTMLCollection` interface added a few days ago in this commit by @nmn https://github.com/facebook/react-strict-dom/commit/2c06e8437b8c367482a72cb11a16d89c4e065f11

What happens is that the Typescript types generated for it are like so:
```ts
export interface StrictHTMLCollection<Elem extends StrictHTMLElement> {
  @@iterator(): Iterator<Elem>;
  readonly length: number;
  readonly [index: number | string]: Elem;
}
```
In `StrictHTMLFormElements.d.ts` - which I can't link because it looks like the TS types are only generated at bundling time for new npm releases(?)

And that `@@iterator(): Iterator<Elem>;` line is not valid TS syntax. It should be `[Symbol.iterator](): IterableIterator<Elem>;`. So probably it's more of a root problem on `flow-api-translator` but until this type is fixed in RSD we can't bump to 0.0.6.

### Expected behavior

Types are correct and tsc builds successfully.

### Steps to reproduce

1. create a new `react-native` project (`npx react-native@latest init RSDBaseProject`)
2. add `import type {StrictHTMLCollection} from 'react-strict-dom/dist/types/StrictHTMLFormElements';` around L20 or so in `App.tsx`
3. `yarn tsc`

### Test case

_No response_

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by nmn**
Fixes #92 

We should fix this issue in `flow-api-translator`, but I've added a patch to the script to fix this for now.

### CODE REVIEW

**Comment by necolas** [comment]
#92 says that the value also needs fixing:

* `@@iterator()` => `[Symbol.iterator]()`
* `Iterator<Elem>` => `IterableIterator<Elem>`

**Comment by necolas** [comment]
When I changed the TS types to match what @kelset said:

```tsx
export interface StrictHTMLCollection<Elem extends StrictHTMLElement> {
  [Symbol.iterator](): IterableIterator<Elem>;
  readonly length: number;
  readonly [index: number | string]: Elem;
}
```

TS error:

```
Property 'length' of type 'number' is not assignable to 'string' index type 'Elem'.ts
(property) StrictHTMLCollection<Elem extends StrictHTMLElement>.length: number
```

Does `index` need to be string or number?

**Comment by nmn** [comment]
> Does index need to be string or number?

Yes, the `number` is actually more important than `string` here.

**Comment by necolas** [comment]
Hmm looks like Flow expects `string` to match the dom libdef though.

```
Cannot cast htmlCollection to StrictHTMLCollection because string [1] is incompatible with
number [2] in the indexer property's key. [incompatible-cast]

     packages/react-strict-dom/src/types/flowtests/is-subtype-of-strict-flowtest.js.flow
       40│ (element as StrictElement);
       41│
       42│ declare var htmlCollection: HTMLCollection<HTMLElement>;
       43│ (htmlCollection as StrictHTMLCollection<HTMLElement>);
       44│ (htmlCollection as StrictHTMLCollection<StrictHTMLElement>);
       45│
       46│ // HTMLElement extends Element

     /private/tmp/flow/flowlib_a666e7d7ea1bb41b_501/dom.js
 [1] 1084│   [index: number | string]: Elem;

     packages/react-strict-dom/src/types/StrictHTMLFormElements.js
 [2]   49│   +[index: number]: Elem;
```

**Comment by tido64** [comment]
> When I changed the TS types to match what @kelset said:
> 
> ```tsx
> export interface StrictHTMLCollection<Elem extends StrictHTMLElement> {
>   [Symbol.iterator](): IterableIterator<Elem>;
>   readonly length: number;
>   readonly [index: number | string]: Elem;
> }
> ```
> 
> TS error:
> 
> ```
> Property 'length' of type 'number' is not assignable to 'string' index type 'Elem'.ts
> (property) StrictHTMLCollection<Elem extends StrictHTMLElement>.length: number
> ```
> 
> Does `index` need to be string or number?

What version of TypeScript did you use? We didn't see this error with TypeScript 5.

File: packages/react-strict-dom/src/types/StrictHTMLCollection.d.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow strict
```
[the remark this answers is not in the document]
Comment (nmn) [reply packages/react-strict-dom/src/types/StrictHTMLCollection.d.ts]: Thanks for catching!

**Comment by necolas** [comment]
Why do flow and ts expect different types though? Is it a bug in the Flow dom types?

**Pull request merged**